eras/conway/impl/src/Cardano/Ledger/Conway/State/Stake.hs

```diff
@@ -50,7 +52,15 @@ newtype ConwayInstantStake era = ConwayInstantStake
 
 instance DecShareCBOR (ConwayInstantStake era) where
   type Share (ConwayInstantStake era) = Interns (Credential 'Staking)
-  decShareCBOR credInterns = ConwayInstantStake <$> decShareCBOR (credInterns, mempty)
+  decShareCBOR credInterns = do
+    peekTokenType >>= \case
```

Comment:
Sorry, was gonna ask during the meeting but then I forgot.  I can't figure out how come this gives us backwards compatibility  - specifically what is the meaning of these first 3 cases?  I understand it's to do with pointers - but is it in any way related to the now deceased IncrementalStake ? Or nothing to do with that?  This function is a mystery to me, due to my ignorance -  sorry!

Reply:
It is a very good question. We use similar approach to provide backwards compatibility for TxOut (see Babbage or Alonzo TxOut de-serialization).
The idea is that current serialization for Conway expects a `Map`, but previous serialization had a list of two elements with two maps. So, in order to be able to deserialize previous representation we can check what is the upcoming token and if it indicates serialization from the previous representation (i.e. any token that indicates a list), then we know that it is not the new format, so we use deserializer from the past and discard the stuff that was dropped. Otherwise we proceed with serialization that is expected for the new type, i.e. the single Map.

Hopefully that clarifies this trick a bit.

Reply:
I see, so because of this (and stake encoded  like this): 

```
data IncrementalStake = IStake
  { credMap :: !(Map (Credential 'Staking) (CompactForm Coin))
  , ptrMap :: !(Map Ptr (CompactForm Coin))
  }
  deriving (Generic, Show, Eq, Ord, NoThunks, NFData)

instance EncCBOR IncrementalStake where
  encCBOR (IStake st dangle) =
    encodeListLen 2 <> encCBOR st <> encCBOR dangle
```

That makes sense, thank you for the thorough explanation.